### ISSUE TITLE
Request: support for PHP built-in server

### ISSUE THREAD

**Issue by vaites** [lrsjng/h5ai#571]
PHP 5.4 and greater comes with a basic [built-in web server](http://php.net/manual/en/features.commandline.webserver.php). h5ai doesn't support it but with a very simple router file works without major issues:
```
// full path of requested file
$path = dirname(__DIR__) . preg_replace('/\?(.*)/', '', $_SERVER['REQUEST_URI']);

// h5ai is required when user request a directory
if(file_exists($path) && is_dir($path))
{
    // directory must not contain index.php or index.html
    if(file_exists("$path/index.php") == false && file_exists("$path/index.html") == false)
    {
        // script name must be "tweaked" to detect path
        $_SERVER['SCRIPT_NAME'] = '/_h5ai/public/index.php';

        include __DIR__ . "/public/index.php";
        exit;
    }
}

// CSS is sent with text/html content type, so we need to fix it
if($path == __DIR__ . '/public/css/styles.css')
{
    header('Content-Type: text/css');
}

// return resource "as-is"
return false;
```
If you put this code in a `_h5ai/router.php` and run `php -S localhost:8000 _h5ai/router.php` being in the root of the web server, h5ai works like a charm.

Can this feature be integrated in h5ai? Didn't make a pull request because of don't know how do you want to integrate this code. I think the best way is a standalone router file...

Why use h5ai with built-in web server? Because it can be used to enhance very lightweight servers in embedded systems like OpenWRT or Android without the need of LightHTTPd, Nginx or Apache.

**Comment by lrsjng** [comment]
@vaites that's a nice addition. Tested it with PHP `5.5.21`, `5.6.5` and `7.0.7`. Unfortunately PHP 7 failed. There are some weird bits sent before the HTTP header that scramble all files that are passed through ...

**Comment by RomeSilvanus** [comment]
OSX 10.12, Safari 10 / Chromium 53.0.2768.0 (64-bit)
```
PHP 5.6.24 (cli) (built: Aug  1 2016 19:30:46) 
Copyright (c) 1997-2016 The PHP Group
Zend Engine v2.6.0, Copyright (c) 1998-2016 Zend Technologies
```

Command:
```
romesmbp:_h5ai rome$ php -S localhost:8000 _h5ai/router.php
PHP 5.6.24 Development Server started at Tue Aug 16 14:27:22 2016
Listening on http://localhost:8000
Document root is /Users/rome/Desktop/_h5ai
Press Ctrl-C to quit.
```


Error:
```
Warning: Unknown: failed to open stream: No such file or directory in Unknown on line 0

Fatal error: Unknown: Failed opening required '_h5ai/router.php' (include_path='.:') in Unknown on line 0
```

**Comment by lrsjng** [comment]
@RomeSilvanus run the command from `/Users/rome/Desktop/`

**Comment by vaites** [comment]
Thanks @lrsjng. I tested it with PHP 5.6.24 and PHP 7.0.9 and worked great. I've just compiled PHP 7.0.7 and works well too... What are these weird bits?

**Comment by RomeSilvanus** [comment]
@lrsjng Running it from Desktop onl leads to:
```// full path of requested file $path = dirname(__DIR__) . preg_replace('/\?(.*)/', '', $_SERVER['REQUEST_URI']); // h5ai is required when user request a directory if(file_exists($path) && is_dir($path)) { // directory must not contain index.php or index.html if(file_exists("$path/index.php") == false && file_exists("$path/index.html") == false) { // script name must be "tweaked" to detect path $_SERVER['SCRIPT_NAME'] = '/_h5ai/public/index.php'; include __DIR__ . "/public/index.php"; exit; } } // CSS is sent with text/html content type, so we need to fix it if($path == __DIR__ . '/public/css/styles.css') { header('Content-Type: text/css'); } // return resource "as-is" return false;```

**Comment by lrsjng** [comment]
@RomeSilvanus that's better :) just put a `<?php` on top of the router file..

**Comment by RomeSilvanus** [comment]
@lrsjng I though something was missing. Now it works fine!

**Comment by lrsjng** [comment]
@vaites 
![screenshot from 2016-08-16 14-47-57](https://cloud.githubusercontent.com/assets/181815/17699398/c1226276-63c0-11e6-98af-2d319af7505b.png)

most of these are 0-bytes, some 3-bytes, nothing to obvious..

**Comment by vaites** [comment]
I'm testing against h5ai 0.29, are you using other version?.

**Comment by vaites** [comment]
I tested a few options and the only reason I found for these characters is the router file sending characters before returning false. Happens only with scripts.js with any other file?

**Comment by lrsjng** [comment]
@vaites yes, it's a fresh download of release 0.29.0. All files that get passed through have these bytes. Can't capture them to file so far, as `curl` seems to skip them..

**Comment by vaites** [comment]
The file is the same for each PHP version? I cannot think of anything else, because the example I posted worked for me on 4 different PHP versions (5.5.38, 5.6.24, 7.0.7 and 7.0.9).

**Comment by lrsjng** [comment]
yeah all the same, but I compiled the versions slightly different, maybe that's the reason..

```
./configure --prefix=/path/to/php-7.0.7  --enable-cgi --enable-mbstring --enable-exif --with-gd --with-jpeg-dir=/usr/lib --with-png-dir=/usr/lib --with-zlib-dir=/usr/lib
./configure --prefix=/path/to/php-5.5.21 --enable-cgi                   --enable-exif --with-gd --with-jpeg-dir=/usr/lib --with-png-dir=/usr/lib
./configure --prefix=/path/to/php-5.6.5  --enable-cgi                   --enable-exif --with-gd --with-jpeg-dir=/usr/lib --with-png-dir=/usr/lib
```

**Comment by vaites** [comment]
This is my configure: 
```
./configure --prefix=/opt/php --enable-mbstring --enable-zip --enable-bcmath --enable-pcntl --enable-ftp --enable-exif --enable-calendar --enable-sysvmsg --enable-sysvsem --enable-sysvshm --enable-wddx --with-curl --with-mcrypt --with-iconv --with-gmp --with-pspell --with-gd --with-jpeg-dir=/usr --with-png-dir=/usr --with-zlib-dir=/usr --with-xpm-dir=/usr --with-freetype-dir=/usr --enable-gd-native-ttf --enable-gd-jis-conv --with-openssl --with-pdo-mysql=/usr --with-gettext=/usr --with-zlib=/usr --with-bz2=/usr --with-recode=/usr --with-mysqli=/usr/bin/mysql_config --enable-fpm --with-readline --without-apxs2
```
Will try with yours to see if fails...
```

**Comment by lrsjng** [comment]
I compiled `7.0.9` with both of my prev configs and indeed the on with `mbstring` and `zlib` breaks h5ai again while the other one works fine...

**Comment by vaites** [comment]
I tried with and without mbstring and zlib and did not fail. Can you add the code to test against it? . If you prefer, tell me where you want to place the code and I can make a pull request with my working code...

**Comment by lrsjng** [comment]
this is the router file I'm using (just added `.txt` to the filename, so I can upload it here)
[router.php.txt](https://github.com/lrsjng/h5ai/files/425210/router.php.txt)

**Comment by vaites** [comment]
I tried your file and works well with PHP 7.0.7 and PHP 7.0.9. I think is obvious that something is sending these characters before the router, but can't understand what and how... Do you have an auto_prepend_file in php.ini?. Are you using Mac? (I can test on Mac if is necessary).

**Comment by lrsjng** [comment]
The bytes are caused by the zlib output compression. The PHP server fails to send the right HTTP header here. Adding

```
ini_set('zlib.output_compression', '0');
```

at the top of the router file fixes these issues.

**Comment by vaites** [comment]
Great!. I have zlib disabled (my Fedora uses zlib v1.2.8). Tell me if you need something more to add the feature.

Thanks!

**Comment by stevenspinoza** [comment]
Good stuff, and good looking folder trees. Just as @vaites said "**put/create this code in _h5ai/router.php** and run **php -S localhost:8000 _h5ai/router.php**" on the root, also change any index.php to index1.php for allowing other files being fetched.

**Comment by biast12** [comment]
just tested this and it works perfectly for the latest version of php, i did fall for the no `<?php` at the start tho, @vaites could you maybe edit your first "issue" to include the `<?php` at the start?
